framework/src/main/java/org/checkerframework/common/accumulation/AccumulationAnnotatedTypeFactory.java

```diff
@@ -0,0 +1,296 @@
+            if (returnsThis(tree)) {
+                // There is a @This annotation on the return type of the invoked method.
+
+                // Fetch the current type of the receiver, or top if none exists.
+                ExpressionTree receiverTree = TreeUtils.getReceiverTree(tree.getMethodSelect());
+                AnnotatedTypeMirror receiverType;
+                AnnotationMirror receiverAnno;
+
+                if (receiverTree != null
+                        && (receiverType = getAnnotatedType(receiverTree)) != null) {
+                    receiverAnno = receiverType.getAnnotationInHierarchy(TOP);
+                } else {
+                    receiverAnno = TOP;
+                }
+
+                type.replaceAnnotation(receiverAnno);
+            }
+            return super.visitMethodInvocation(tree, type);
+        }
+    }
+
+    @Override
+    public QualifierHierarchy createQualifierHierarchy(MultiGraphFactory factory) {
+        return new AccumulationQualifierHierarchy(factory);
+    }
+
+    /**
+     * All accumulation analyses share a similar type hierarchy. This hierarchy implements the
+     * subtyping, LUB, and GLB for that hierarchy. The lattice looks like:
+     *
+     * <pre>
+     *       top
+     *      /   \
+     * acc(x)   acc(y) ...
+     *      \   /
+     *     acc(x,y) ...
+     *        |
+     *      bottom
+     * </pre>
+     */
+    protected class AccumulationQualifierHierarchy extends MultiGraphQualifierHierarchy {
+
+        /**
+         * Create the qualifier hierarchy
+         *
+         * @param factory the factory
+         */
+        public AccumulationQualifierHierarchy(MultiGraphFactory factory) {
+            super(factory);
+        }
+
+        @Override
+        public AnnotationMirror getTopAnnotation(final AnnotationMirror start) {
+            return TOP;
+        }
+
+        /**
+         * GLB in this type system is set union of the arguments of the two annotations, unless one
+         * of them is bottom, in which case the result is also bottom.
+         */
+        @Override
+        public AnnotationMirror greatestLowerBound(
+                final AnnotationMirror a1, final AnnotationMirror a2) {
+            if (AnnotationUtils.areSame(a1, BOTTOM) || AnnotationUtils.areSame(a2, BOTTOM)) {
+                return BOTTOM;
+            }
+
+            if (!AnnotationUtils.hasElementValue(a1, "value")) {
+                return a2;
+            }
+
+            if (!AnnotationUtils.hasElementValue(a2, "value")) {
+                return a1;
+            }
+
+            if (isAccumulatorAnnotation(a1) && isAccumulatorAnnotation(a2)) {
+                Set<String> a1Val =
+                        new LinkedHashSet<>(
+                                ValueCheckerUtils.getValueOfAnnotationWithStringArgument(a1));
+                Set<String> a2Val =
+                        new LinkedHashSet<>(
```

Comment:
To follow up on this, I just looked through how the Value Checker handles subtyping, etc. between `@StringVal` annotations, and I'm going to do the same thing here. It uses `List` rather than converting to sets in the `QualifierHierarchy`, and sorts and removes at the same time by converting to a `TreeSet` and back in the method that produces new `StringVal` annotations.